src/Common/src/TypeSystem/Common/TypeDesc.cs

```diff
@@ -465,5 +481,25 @@ public virtual MethodDesc GetFinalizer()
         {
             return null;
         }
+
+        /// <summary>
+        /// Gets a value indicating whether this type has generic variance (the definition of the type
+        /// has a generic parameter that is co- or contravariant).
+        /// </summary>
+        public bool HasVariance
+        {
+            get
+            {
+                return (GetTypeFlags(TypeFlags.HasGenericVariance | TypeFlags.HasGenericVarianceComputed) & TypeFlags.HasGenericVariance) != 0;
+            }
+        }
+
+        /// <summary>
+        /// Gets the kind of this type.
+        /// </summary>
+        public abstract TypeKind Variety
```

Comment:
That ended up looking so different from what's in NUTC that I started to doubt I was getting equivalent behaviors and stopped mid-work. I would prefer we don't introduce a third casting algorithm.

I made `Variety` local to the casting code and pulled it out of the rest of the system.

If we decide we want to port the casting from `src\Runtime.Base\src\System\Runtime\TypeCast.cs` later, I don't have any emotional attachment to this code. We can just do it when it's needed. The code will still be different in subtle ways (the handling of generic definitions and RuntimeDeterminedTypes I already mentioned, handling of array enumerators, etc.).